clippy_lints/src/methods/ref_mut_iter_method_chain.rs

```diff
@@ -0,0 +1,40 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::in_macro;
+use clippy_utils::source::snippet_with_context;
+use clippy_utils::ty::implements_trait;
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::{BorrowKind, Expr, ExprKind, Mutability, UnOp};
+use rustc_lint::LateContext;
+use rustc_span::sym;
+
+use super::REF_MUT_ITER_METHOD_CHAIN;
+
+pub(crate) fn check(cx: &LateContext<'_>, self_arg: &Expr<'_>) {
+    if_chain! {
+        if let ExprKind::AddrOf(BorrowKind::Ref, Mutability::Mut, base_expr) = self_arg.kind;
+        if !in_macro(self_arg.span);
+        if let Some(&iter_trait) = cx.tcx.all_diagnostic_items(()).get(&sym::Iterator);
+        if implements_trait(cx, cx.typeck_results().expr_ty(base_expr).peel_refs(), iter_trait, &[]);
+        then {
+            let snip_span = match base_expr.kind {
+                ExprKind::Unary(UnOp::Deref, e) if cx.typeck_results().expr_ty(e).is_ref() && !in_macro(base_expr.span)
```

Comment:
I assume this is to provide the correct suggestion in the last test case? What happens if something is dereferenced twice?

Reply:
It's to make a nicer suggestion on a re-borrow. e.g.

```rust
let x = &mut iter;
(&mut *iter).next();
```

would become

```rust
let x = &mut iter;
iter.by_ref().next();
```

It also does happen to fix a precedence issue with prefix operators before a function call. Multiple can safely be removed so long as they don't call a custom deref implementation.

Reply:
The precedence issue is fixed now. It will keep the parenthesis if they're needed. This defeats the stated purpose of the lint, but it's an edge case anyways.